controller/test/frame_detector/frame_detector_test.cpp

```diff
@@ -0,0 +1,236 @@
+  EXPECT_FALSE(frame_detector.is_frame_available());
+
+  for (uint32_t i = 0; i < LEN - 1; i++) {
+    rx_buf.PutByte(' ');
+  }
+  rx_buf.PutByte(MARK);
+
+  EXPECT_EQ(State::RECEIVING_FRAME, frame_detector.get_state());
+  ASSERT_FALSE(frame_detector.is_frame_available());
+}
+
+TEST(FrameDetector, ErrorWhileRx) {
+  FrameDetectorTest frame_detector(rx_buf);
+  EXPECT_TRUE(frame_detector.Begin());
+  ASSERT_EQ(State::LOST, frame_detector.get_state());
+  EXPECT_FALSE(frame_detector.is_frame_available());
+
+  rx_buf.PutByte(MARK);
+  ASSERT_EQ(State::RECEIVING_FRAME, frame_detector.get_state());
+  EXPECT_FALSE(frame_detector.is_frame_available());
+
+  frame_detector.OnRxError(RxError::TIMEOUT);
+  ASSERT_EQ(State::LOST, frame_detector.get_state());
+  EXPECT_FALSE(frame_detector.is_frame_available());
+
+  rx_buf.PutByte(MARK);
+  ASSERT_EQ(State::RECEIVING_FRAME, frame_detector.get_state());
+  EXPECT_FALSE(frame_detector.is_frame_available());
+
+  frame_detector.OnRxError(RxError::DMA);
+  ASSERT_EQ(State::LOST, frame_detector.get_state());
+  EXPECT_FALSE(frame_detector.is_frame_available());
+
+  rx_buf.PutByte(MARK);
+  ASSERT_EQ(State::RECEIVING_FRAME, frame_detector.get_state());
+  EXPECT_FALSE(frame_detector.is_frame_available());
+
+  frame_detector.OnRxError(RxError::OVERFLOW);
+  ASSERT_EQ(State::LOST, frame_detector.get_state());
+  EXPECT_FALSE(frame_detector.is_frame_available());
+
+  rx_buf.PutByte(MARK);
+  ASSERT_EQ(State::RECEIVING_FRAME, frame_detector.get_state());
+  EXPECT_FALSE(frame_detector.is_frame_available());
+
+  frame_detector.OnRxError(RxError::SERIAL_FRAMING);
+  ASSERT_EQ(State::LOST, frame_detector.get_state());
+  EXPECT_FALSE(frame_detector.is_frame_available());
+
+  rx_buf.PutByte(MARK);
+  ASSERT_EQ(State::RECEIVING_FRAME, frame_detector.get_state());
+  EXPECT_FALSE(frame_detector.is_frame_available());
+
+  frame_detector.OnRxError(RxError::UNKNOWN);
+  ASSERT_EQ(State::LOST, frame_detector.get_state());
+  EXPECT_FALSE(frame_detector.is_frame_available());
+}
+
+vector<string> fakeRx(const char *frame, FrameDetectorTest &frame_detector) {
+  vector<string> ret;
+  for (uint32_t i = 0; i < strlen(frame); i++) {
+    rx_buf.PutByte(frame[i]);
+    if (frame_detector.is_frame_available()) {
+      string s;
+      s.assign(reinterpret_cast<char *>(frame_detector.TakeFrame()),
+               static_cast<size_t>(frame_detector.get_frame_length()));
+      ret.push_back(s);
+    }
+  }
+  return ret;
+}
+
+TEST(FrameDetector, FuzzMoreMarkers) {
+  FrameDetectorTest frame_detector(rx_buf);
+  EXPECT_TRUE(frame_detector.Begin());
+  ASSERT_EQ(State::LOST, frame_detector.get_state());
+  EXPECT_FALSE(frame_detector.is_frame_available());
+
+  vector<string> results = fakeRx(".aaa.", frame_detector);
+  EXPECT_EQ(static_cast<unsigned int>(1), results.size());
+  EXPECT_EQ(0, results[0].compare("aaa"));
+
+  EXPECT_TRUE(frame_detector.Begin());
```

Comment:
In all of these sequences you're basically resetting the FrameDetector. Seems like you could make things much easier by changing FakeRx to create a fresh FrameDetector internally. Then all of these sequences of lines, e.g. 209-214, 216-221 etc., each become a single line.

Reply:
Getting there, sorry for all the copy paste from the last time, I had very little time and did things hastily.